### ISSUE TITLE
Write some tests

### ISSUE THREAD

**Issue by ChShersh** [kowainik/co-log#22]
Tests can be implemented on top of the existing `PureLoggerT` monad

**Comment by ChShersh** [comment]
@arkrost That would be great! The issue is still relevant :slightly_smiling_face: Feel free to ask any questions!

**Comment by arkrost** [comment]
I'm expected to use Tasty, hspec and hedgehog, aren't I? Just briefly looked through other projects.

**Comment by ChShersh** [comment]
Earlier at work we switched from `tasty` + `tasty-discover` + `hspec` + `hedgehog` to the same stack of testing tools but without `tasty-discover` because we needed more customization. And just recently we get rid of `tasty` and currently using `hspec` for unit testing and `hedgehog` for property-based tests. But it's okay if you can take other packages as examples and implement tests with `tasty` if that is easier for you.

I think `hedgehog` can be used to check typeclasses laws, like `(x <> y) <> z = x <> (y <> z)`. But it's not a trivial task because `LogAction` is a monadic function and testing such functions for equality is not that easy... 

Another possibility to use property-based tests is it use _command_ technique to introduce ADT with possible commands for log actions, generate random list of such commands, implement interpreter for the list of such commands and then implement verification functions for the result of logging.

Regarding unit tests: I assume that `PureLogger` monad will be used for testing because testing IO logger actions is also not that easy. With pure monad it's much easier to verify result of logging action.

**Comment by arkrost** [comment]
I'm stuck with the necessity to explicitly write return types for expressions like
``` haskell
runPureLog $ usingLoggerT logMessagePure $ logMsg "aaa"
```
it fails with the following error
``` haskell
    • Could not deduce (HasLog
                          (LogAction (LoggerT msg (PureLoggerT msg Identity)) msg)
                          msg0
                          (LoggerT msg (PureLoggerT msg Identity)))
      from the context: (HasLog
                           (LogAction (LoggerT msg (PureLoggerT msg Identity)) msg)
                           msg1
                           (LoggerT msg (PureLoggerT msg Identity)),
                         IsString msg1)
        bound by the inferred type for ‘it’:
                   forall msg msg1.
                   (HasLog
                      (LogAction (LoggerT msg (PureLoggerT msg Identity)) msg)
                      msg1
                      (LoggerT msg (PureLoggerT msg Identity)),
                    IsString msg1) =>
                   ((), [msg])
        at <interactive>:5:1-55
      The type variable ‘msg0’ is ambiguous
    • In the ambiguity check for the inferred type for ‘it’
      To defer the ambiguity check to use sites, enable AllowAmbiguousTypes
      When checking the inferred type
        it :: forall msg1 msg2.
              (HasLog
                 (LogAction (LoggerT msg1 (PureLoggerT msg1 Identity)) msg1)
                 msg2
                 (LoggerT msg1 (PureLoggerT msg1 Identity)),
               IsString msg2) =>
              ((), [msg1])
```

Could you please explain the roots of the problem?

**Comment by ChShersh** [comment]
@arkrost Does the #65 resolves your issue?

**Comment by arkrost** [comment]
@ChShersh No, #65 is not related to my problem.

**Comment by ChShersh** [comment]
@arkrost From the error it looks like compiler cannot infer type properly and infers more polymorphic type. Try to do something like that:

```haskell
example :: WithLog env String m => m ()
example = logMsg "aaa"
```

and then

```haskell
runPureLog $ usingLoggerT logMessagePure example
```

I'm mostly looking at this piece of code:

https://github.com/kowainik/co-log/blob/4ae3f06df8a4d2da20c9e12c107204857e3cda88/co-log/example/Main.hs#L121-L124

**Comment by arkrost** [comment]
It seems I won't be able to return this task in the following 3 weeks. So if someone want to pick this one, feel free to take it.

**Comment by ChShersh** [comment]
Resolved in #67
